api/v1beta1/glance_types.go

```diff
@@ -0,0 +1,182 @@
+	ServiceUser string `json:"serviceUser"`
+
+	// +kubebuilder:validation:Required
+	// Glance Container Image URL
+	ContainerImage string `json:"containerImage,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// MariaDB instance name
+	// Right now required by the maridb-operator to get the credentials from the instance to create the DB
+	// Might not be required in future
+	DatabaseInstance string `json:"databaseInstance,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=glance
+	// DatabaseUser - optional username used for glance DB, defaults to glance
+	// TODO: -> implement needs work in mariadb-operator, right now only glance
+	DatabaseUser string `json:"databaseUser"`
+
+	// +kubebuilder:validation:Required
+	// Secret containing OpenStack password information for glance GlanceDatabasePassword
+	Secret string `json:"secret"`
+
+	// +kubebuilder:validation:Optional
+	// PasswordSelectors - Selectors to identify the DB password from the Secret
+	PasswordSelectors PasswordSelector `json:"passwordSelectors,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Debug - enable debug for different deploy stages. If an init container is used, it runs and the
+	// actual action pod gets started with sleep infinity
+	Debug GlanceDebug `json:"debug,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// PreserveJobs - do not delete jobs after they finished e.g. to check logs
+	PreserveJobs bool `json:"preserveJobs,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="# add your customization here"
+	// CustomServiceConfig - customize the service config using this parameter to change service defaults,
+	// or overwrite rendered information using raw OpenStack config format. The content gets added to
+	// to /etc/<service>/<service>.conf.d directory as custom.conf file.
+	CustomServiceConfig string `json:"customServiceConfig,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// ConfigOverwrite - interface to overwrite default config files like e.g. logging.conf or policy.json.
+	// But can also be used to add additional files. Those get added to the service config dir in /etc/<service> .
+	// TODO: -> implement
+	DefaultConfigOverwrite map[string]string `json:"defaultConfigOverwrite,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// StorageClass
+	StorageClass string `json:"storageClass,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// StorageRequest
+	StorageRequest string `json:"storageRequest"`
+
+	// +kubebuilder:validation:Required
+	// GlanceAPIInternal - Spec definition for the internal and admin API service of this Glance deployment
+	GlanceAPIInternal GlanceAPISpec `json:"glanceAPIInternal"`
+
+	// +kubebuilder:validation:Required
+	// GlanceAPIExternal - Spec definition for the external API service of this Glance deployment
+	GlanceAPIExternal GlanceAPISpec `json:"glanceAPIExternal"`
+}
```

Comment:
Will need to ask the storage folks about this

Reply:
@fmount I actually kept the `CephBackend` in the `GlanceAPI` and did not move it up into the top-level `Glance` CRD.  Since that seems to be what you suggest we ultimately want, it looks like this might be okay for now?

Reply:
yes, looks good

Reply:
>... However, regarless of the endpoint config, different `GlanceAPI` instances might have different Ceph clusters backend (with diff info), ...

you mean we would run a glance-api per ceph cluster and not one api with multiple ceph backends? how would that map to the glance-operator if you have different Ceph clusters and split this in multiple api instances? do we have multiple top level Glance CRDs which then creates an external/internal glance-api instance for each ceph cluster, or should the glanceapi in the top level CRD just be a list of glanceapis with one external and multiple internal apis for the different ceph clusters?

Reply:
@abays @stuggi  I was thinking more on this change, and I think having a `glance-api` for each ceph cluster is not the right approach. Considering that the current implementation does not support `multiple ceph clusters` for a single API instance, you can potentially just have a diff backend on internal and external, but I suspect this is not the goal here and it results on an unexpected behavior from the user perspective. 
Instead, we want to propagate the cluster information to the multiple `glance-api` instances, and then enable or not the backend injecting a specific config (via `CustomServiceConfig`).
In cinder we're relying on a global `CephBackend` structure definition atm, which is not flexible, but we're working on a new proposal to switch the implementation on having multiple secrets instead of Backend definition.
The are several drawbacks on the current approach:

1. The operator CRD is too verbose (and it would be more verbose having a `CephBackend` per `glance-api` instance) and it’s aware about several details of the cluster;
2. Part of this information (`ClientKey`, `ExternalMonIpAddresses`, `RBDUsers`) are stored as plaintext within `ConfigMaps`, and this raise security concerns;
3. Everything is stored within the __same__ `ConfigMap`: there’s no clear distinction between the Ceph data configuration and the operator specifics;
4. The main operator controller is responsible to render the Ceph config files using the information passed as input: onboard processing means that implementation details should be available at operator level;
5. Having high coupled Ceph data and operator config increase the chances to see crashes in the operator due to wrong parameters related to the ceph cluster definition

Being that said, and since we don't want to block the current PR, @abays wdyt about moving the `CephBackend` definition to the top level crd (and `glance-api` instances can just pull it down from the parent crd).
This would simplify the current usage of the glance operator and keep it in sync with cinder, until the new approach is approved (and part of a follow up PR).